### PR TITLE
Report unhappy blocks in happy path test

### PR DESCRIPTION
**Pull request by al8n**
As mentioned in the meeting, we want to make sure in happy path tests, if a block contains AggratedQc, then we need to persist the block information.

I implemented a `drop` method for `Block` and feature gate it, this way we do not need to modify any of our current codebase, and we can easily remove the code and feature in the future when we fully solve the problem.

### CODE REVIEW

File: nomos-core/src/block/debug_happy.rs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+            let info = Info {
+                id: format!("{}", header.id),
+                view: header.view,
+                qcs,
+            };
+            // Use pretty print to make it easier to read, because we need the this for debugging.
+            serde_json::to_writer_pretty(&mut *file, &info).unwrap();
+        }
+    }
+}
```
Comment (danielSanchezQ) [inline nomos-core/src/block/debug_happy.rs:35]: I think this should be a simple method or function and not do it in drop. It is better to see the call explicetely it in the test.

**Review by danielSanchezQ — CHANGES_REQUESTED** [review]

**Review by Zeegomo — CHANGES_REQUESTED** [review]
Agree with Daniel, but apart from that don't we already have the `safe_blocks` field in https://github.com/logos-co/nomos-node/blob/2ceba073c060476fb509d20f7a8515962b173c34/nomos-services/consensus/src/lib.rs#L926

**Comment by al8n** [comment]
> Agree with Daniel, but apart from that don't we already have the `safe_blocks` field in
> 
> https://github.com/logos-co/nomos-node/blob/2ceba073c060476fb509d20f7a8515962b173c34/nomos-services/consensus/src/lib.rs#L926
> 
> ?

Yeah, we have `safe_blocks`, I previously thought we wanted to do the check for each block. Now, just use `safe_blocks`.

File: tests/src/tests/happy.rs [hunk 2]
```diff
@@ -1,13 +1,44 @@
...
+            .map(|b| Info {
+                id: format!("{}", header.id),
+                view: header.view,
+                qcs,
+            })
+            .collect::<Vec<_>>();
+        // Use pretty print to make it easier to read, because we need the this for debugging.
+        serde_json::to_writer_pretty(&mut *file, &infos).unwrap();
+    }
+}
```
Comment (danielSanchezQ) [inline tests/src/tests/happy.rs:39]: I don't think we should persist them, here are the steps I think need to be done:

* We can add a query for requesting the blocks to the node.
* Before killing the nodes, retrieve blocks.
* Check blocks with this function (or similar).
* Report `NodeId` `BlockId` `View` for every block that is not starndardQc

Comment (Zeegomo) [reply tests/src/tests/happy.rs:39]: We already have an api to get blocks from the node, I think we can reuse that. In fact, it's the same that we use to check the view, it also has an additional field with blocks info

Comment (danielSanchezQ) [reply tests/src/tests/happy.rs:39]: > We already have an api to get blocks from the node, I think we can reuse that. In fact, it's the same that we use to check the view, it also has an additional field with blocks info

Even better!

**Comment by al8n** [comment]
> > I previously thought we wanted to do the check for each block
> 
> What do you mean? `safe_blocks` contain all blocks, so we can check that all of them have an happy qc

I 

> > I previously thought we wanted to do the check for each block
> 
> What do you mean? `safe_blocks` contain all blocks, so we can check that all of them have an happy qc

Ah, I think there may be some "intermediate" blocks, and we also want to check it, so I just implemented a drop method for it .

**Comment by Zeegomo** [comment]
> Ah, I think there may be some "intermediate" blocks, and we also want to check it

All intermediate blocks are in safe blocks apart from outright malicious / invalid ones

**Review by bacv — APPROVED** [review]

**Review by danielSanchezQ — CHANGES_REQUESTED** [review]
Just missing the nodes id that have the unhappy block in the info. Otherwise this is it!!

**Review by youngjoon-lee — APPROVED** [review]
Looks good! + I agree with https://github.com/logos-co/nomos-node/pull/430#pullrequestreview-1637460245

**Review by Zeegomo — APPROVED** [review]

**Pull request merged**